pkg/bb/bbmain/cmd/main.go

```diff
@@ -27,7 +28,96 @@ func main() {
+// files but specifying bb as the interpreter.
+//
+// ls would then be:
+// #!/bin/bb ls
+//
+// Note that the absolute path is required, else Linux will throw an error as bb
+// is not in the list of allowed interpreters.
+// The /bin/bb path is not an issue on Plan 9, since users construct their name space
+// on startup and binding /bbin into /bin is no problem.
+//
+// In this case the kernel will stat, open, and read the file, find the executable name,
+// and start it. This approach has as low overhead as the symlink approach.
+//
+// One problem remains: Unix and Plan 9 evaluate arguments in a #! file differently,
+// and, further, invoke the argument in a different way.
+// Given the file shown above, bb on Linux gets the arguments:
+// [ls ls /tmp/ls]
+// With the same file, bb on linux gets this:
+// [/bbin/bb ls /tmp/ls]
+// But wait! There's more!
+// On Plan 9, the arguments following the interpreter are tokenized (split on space)
+// and on Linux, they are not.
+//
+// This leads to a few conclusions:
+// - We can get around lack of symlinks by using #! (sh-bang) files with an absolute path to
+//   bb as the interpreter, e.g. #!/abs/path/to/bb argument.
+//   This achieves the "exec once" goal.
+// - We can specify which u-root tool to use via arguments to bb in the #! file.
+// - The argument to the interpreter (/bbin/bb) should be one token (e.g. ls) because of different
+//   behavior in different systems (some tokenize, some do not).
+// - Because of the differences in how arguments are presented to #! on different kernels,
+//   there should be a reasonably unique marker so that bb can have confidence that
+//   it is running as an interpreter.
+//
+// The conclusions lead to the following design:
+// #! files for bb specify their argument with #!. E.g., the file for ls looks like this:
+// #!/bbin/bb #!ls
+// On Linux, the args to bb then look like:
+// [/bbin/bb #!ls /tmp/ls ...]
+// on Plan 9:
+// [ls #!ls /tmp/ls ...]
+// The code needs to change the arguments to look like an exec:
+// [/tmp/ls ...]
+// In each case, the second arg begins with a #!, which is extremely unlikely to appear
+// in any other context (save testing #! files, of course).
+// The result is that the kernel, given a path to a u-root #! file, will read that file,
+// then exec bbin with the argument from the #! and any additional arguments from the exec.
+// The overhead in this case is no more than the symlink overhead.
+// A final advantage is that we can now install u-root on file systems that don't have
+// symbolic links, e.g. VFAT, and it will have low overhead.
+//
+// So, dear reader, if you are wondering why the little bit of code below is the way
+// it is, now you know.
 func init() {
+	// If this has been run from a #! file, it will have at least
+	// 3 args, and os.Args needs to be reconstructed.
+	if len(os.Args) > 2 && strings.HasPrefix(os.Args[1], "#!") {
+		os.Args = []string{os.Args[2]}
```

Comment:
Should this be assigning to a temp var?  How can you assign an array of length 1 and then compare len > 3 in the next line?

Am I missing something?

Reply:
yes you can, but chris pointed out that it can be simpler, I'm doing tha tnow. But the code works as is.